### PR TITLE
Validated implementation of the AWS Load Balancer Controller and docu…

### PR DESCRIPTION
**Pull request by shapirov103**
Validated implementation of the AWS Load Balancer Controller and documentation. Minor fixes

*Issue #, if available:*: related to #70

*Description of changes:*

Adds AWS LB controller implementation and documentation. This is needed for proper NGINX architecture as well as many other use cases, including UDP ingress with NLB, other ingress controllers such as Kong, AWS Native ingress support with ALB. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/pipeline-stack/index.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ import * as ssp from '../../lib'
 import * as team from '../teams'
 
 export default class PipelineStack extends cdk.Stack {
-    constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
+    constructor(scope: cdk.App, id: string, _props?: cdk.StackProps) {
```
Comment (kcoleman731) [inline examples/pipeline-stack/index.ts:13]: whats up with _props?

File: lib/addons/aws-loadbalancer-controller/index.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * Configuration options for AWS Load Balancer controller
+ */
+export interface AwsLoadBalancerControllerProps {
+    /**
+     * Namespace where controller will be installed
+     */
+    namespace?: string,
+    /**
```
Comment (kcoleman731) [inline lib/addons/aws-loadbalancer-controller/index.ts:14]: space before next doc bloc

File: lib/addons/aws-loadbalancer-controller/index.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+const AWS_LOAD_BALANCER_CONTROLLER = 'aws-load-balancer-controller';
+
+export class AwsLoadBalancerControllerAddOn implements ClusterAddOn {
+
+    private options: AwsLoadBalancerControllerProps;
+
+    constructor(props?: AwsLoadBalancerControllerProps) {
+        this.options = { ...AwsLoadBalancerControllerDefaults, ...props ?? {}};
```
Comment (kcoleman731) [inline lib/addons/aws-loadbalancer-controller/index.ts:53]: Don't think a conditional is needed here?

File: lib/addons/aws-loadbalancer-controller/index.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    deploy(clusterInfo: ClusterInfo): void {
+
+        const cluster = clusterInfo.cluster; 
+        const awsControllerBaseResourceBaseUrl = `https://raw.githubusercontent.com/kubernetes-sigs/aws-load-balancer-controller/${this.options.version}/docs`;
+        const awsControllerPolicyUrl = `${awsControllerBaseResourceBaseUrl}/install/iam_policy${cluster.stack.region.startsWith('cn-') ? '_cn' : ''}.json`;
```
Comment (kcoleman731) [inline lib/addons/aws-loadbalancer-controller/index.ts:60]: nit...but move the url vars below service account creation. Makes it clear where the vars are used.

File: lib/addons/aws-loadbalancer-controller/index.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        
+        ((JSON.parse(policyJson)).Statement as []).forEach((statement) => {
+            serviceAccount.addToPrincipalPolicy(iam.PolicyStatement.fromJson(statement));
+        });
+
+        const awsLoadBalancerControllerChart = cluster.addHelmChart('AWSLoadBalancerController', {
+            chart: AWS_LOAD_BALANCER_CONTROLLER,
+            repository: 'https://aws.github.io/eks-charts',
+            namespace: this.options.namespace,
+            release: 'aws-load-balancer-controller',
```
Comment (kcoleman731) [inline lib/addons/aws-loadbalancer-controller/index.ts:77]: use `AWS_LOAD_BALANCER_CONTROLLER` constant?

File: lib/addons/aws-loadbalancer-controller/index.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+        const policyJson = request('GET', awsControllerPolicyUrl).getBody().toString();
+        
+        ((JSON.parse(policyJson)).Statement as []).forEach((statement) => {
+            serviceAccount.addToPrincipalPolicy(iam.PolicyStatement.fromJson(statement));
+        });
+
+        const awsLoadBalancerControllerChart = cluster.addHelmChart('AWSLoadBalancerController', {
+            chart: AWS_LOAD_BALANCER_CONTROLLER,
+            repository: 'https://aws.github.io/eks-charts',
```
Comment (kcoleman731) [inline lib/addons/aws-loadbalancer-controller/index.ts:75]: we should make a constant for this somewhere in the project.

File: lib/addons/aws-loadbalancer-controller/index.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            namespace: this.options.namespace,
+            release: 'aws-load-balancer-controller',
+            version: '1.2.0', // mapping to v2.2.0
```
Comment (kcoleman731) [inline lib/addons/aws-loadbalancer-controller/index.ts:78]: what happens if the version the customers supplies doesn't match?

**Review by kcoleman731 — COMMENTED** [review]
Awesome stuff. A few small comments.

File: examples/pipeline-stack/index.ts (hunk 1 above)
Comment (shapirov103) [reply examples/pipeline-stack/index.ts:13]: Linting and IDE complains that props are never used. But it is important to keep them in the signature. So _ before property name is a common convention for optional properties and at lease VSCode automatically understands it and does not highlight this as an issue. I was thinking may be as a next step could fix linting to drop the warning for such cases as well.

File: lib/addons/aws-loadbalancer-controller/index.ts (hunk 2 above)
Comment (shapirov103) [reply lib/addons/aws-loadbalancer-controller/index.ts:14]: Yep, missed.

File: lib/addons/aws-loadbalancer-controller/index.ts (hunk 3 above)
Comment (shapirov103) [reply lib/addons/aws-loadbalancer-controller/index.ts:53]: Yes, keep forgetting it is ts and a lot less restrictive. Will drop.

File: lib/addons/aws-loadbalancer-controller/index.ts (hunk 4 above)
Comment (shapirov103) [reply lib/addons/aws-loadbalancer-controller/index.ts:60]: Yeah, I stole this one, thought it was pretty good effort to handle cn regions as well.

File: lib/addons/aws-loadbalancer-controller/index.ts (hunk 6 above)
Comment (shapirov103) [reply lib/addons/aws-loadbalancer-controller/index.ts:75]: sounds like a good refactoring task, will affect a lot of things. Let's create the task, I will keep it consistent for now. Agree overall.

File: lib/addons/aws-loadbalancer-controller/index.ts (hunk 7 above)
Comment (shapirov103) [reply lib/addons/aws-loadbalancer-controller/index.ts:78]: It is a miss. I will put this a new prop for the chart version. At this point it is hard to predict how those will map in the future.

File: lib/addons/aws-loadbalancer-controller/index.ts (hunk 5 above)
Comment (shapirov103) [reply lib/addons/aws-loadbalancer-controller/index.ts]: Addressed

**Review by askulkarni2 — APPROVED** [review]
:shipit:

**Pull request merged**